test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c

```diff
@@ -337,11 +352,20 @@ ClassPrepare(jvmtiEnv *jvmti, JNIEnv *env, jthread thread, jclass klass) {
     jvmtiPhase phase;
     intptr_t exp_val = 777;
     intptr_t act_val;
+    jvmtiError err;
+
+    // Block VMDeath event and other ClassPrepare events while this callback is executed.
+    // Sync with VMDeath event guards agains JVMTI_ERROR WRONG_PHASE.
+    err = (*jvmti)->RawMonitorEnter(jvmti, event_mon);
```

Comment:
I don't see how this fixes the problem. You can be exiting VMDeath when this monitor is entered. This does not block the get_thread_local() call below, which is what leads to the WRONG_PHASE. I think using a raw monitor is the right approach, but I was expecting to see VMDeath set a flag after grabbing the monitor that indicated VMDeath had been called. Then ClassPrepare should check this flag after grabbing the monitor and ignore the event if it is set.

Reply:
Thank you for the comment.
Fixed the `get_thread_local()` issue and corrected phase check.
New flag is not really needed as the `phase == JVMTI_PHASE_DEAD` can play its role.
Please, let me know if you are not okay with that.